engines/trecision/3d.cpp

```diff
@@ -0,0 +1,2687 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "trecision/3d.h"
+#include "trecision/anim.h"
+#include "trecision/sound.h"
+#include "trecision/text.h"
+#include "trecision/actor.h"
+#include "trecision/graphics.h"
+#include "trecision/trecision.h"
+#include "trecision/video.h"
+
+namespace Trecision {
+#define SHADOWVERTSNUM 42
+
+static const int16 _shadowVerts[SHADOWVERTSNUM] = {
+	6, 15, 23,
+	24, 32, 78,
+	80, 81, 83,
+	86, 90, 99,
+	107, 108, 116,
+	155, 157, 158,
+	160, 164, 168,
+	169, 173, 174,
+	187, 188, 192,
+	193, 213, 215,
+	227, 229, 235,
+	238, 249, 250,
+	252, 253, 299,
+	306, 330, 336
+};
+#define SHADOWFACESNUM 48
+
+const int16 _shadowFaces[SHADOWFACESNUM][3] = {
+	{22, 21, 5}, {7, 5, 22},
+	{7, 19, 5}, {5, 2, 19},
+	{27, 24, 16}, {27, 16, 18},
+	{18, 16, 9}, {18, 13, 9},
+	{13, 9, 2}, {3, 19, 12},
+	{25, 26, 17}, {17, 15, 25},
+	{17, 19, 15}, {15, 12, 19},
+	{20, 23, 8}, {8, 6, 20},
+	{6, 9, 3}, {3, 8, 6},
+	{12, 3, 4}, {4, 11, 12},
+	{35, 4, 11}, {13, 2, 1},
+	{1, 14, 13}, {14, 37, 1},
+	{1, 34, 37}, {31, 36, 37},
+	{37, 30, 31}, {29, 34, 35},
+	{35, 29, 28}, {36, 11, 31},
+	{30, 37, 14}, {29, 1, 34},
+	{28, 4, 35}, {36, 10, 35},
+	{35, 32, 10}, {37, 0, 34},
+	{37, 33, 0}, {0, 33, 39},
+	{39, 40, 0}, {10, 38, 32},
+	{32, 41, 38}, {36, 35, 34},
+	{36, 37, 35}, {11, 36, 35},
+	{38, 40, 41}, {41, 38, 39},
+	{2, 19, 13}, {3, 9, 12}
+};
+
+Renderer3D::Renderer3D(TrecisionEngine *vm) : _vm(vm) {
+	_zBuffer = new int16[ZBUFFERSIZE / 2];
+
+	_minXClip = 0;
+	_minYClip = 0;
+	_maxXClip = 0;
+	_maxYClip = 0;
+	_zBufStartX = 0;
+	_zBufStartY = 0;
+	_zBufWid = 0;
+	_shadowLightNum = 0;
+	_totalShadowVerts = 0;
+
+	// data for the triangle routines
+	for (int i = 0; i < 480; ++i) {
+		_lEdge[i] = 0;
+		_rEdge[i] = 0;
+		_lColor[i] = 0;
+		_rColor[i] = 0;
+		_lZ[i] = 0;
+		_rZ[i] = 0;
+		_lTextX[i] = 0;
+		_rTextX[i] = 0;
+		_lTextY[i] = 0;
+		_rTextY[i] = 0;
+	}
+
+	for (int i = 0; i < 10; ++i)
+		_shadowIntens[i] = 0;
+
+	for (int i = 0; i < MAXVERTEX; ++i) {
+		_vVertex[i].clear();
+		_shVertex[i].clear();
+	}
+
+	drawObj.rect = Common::Rect(0, 0, 0, 0);
+	drawObj.l = Common::Rect(0, 0, 0, 0);
+}
+
+Renderer3D::~Renderer3D() {
+	delete[] _zBuffer;
+}
+
+void Renderer3D::textureTriangle(int32 x1, int32 y1, int32 z1, int32 c1, int32 tx1, int32 ty1,
+								 int32 x2, int32 y2, int32 z2, int32 c2, int32 tx2, int32 ty2,
+								 int32 x3, int32 y3, int32 z3, int32 c3, int32 tx3, int32 ty3,
+								 STexture *t) {
+	int32 cl;  // color of left edge of horizontal scanline
+	int32 zl;  // zbuffer of left edge of horizontal scanline
+	int32 olx; // texture x of left edge of horizontal scanline
+	int32 oly; // texture y of left edge of horizontal scanline
+	int16 y;   // looping variable
+
+	if (y1 > _maxYClip)
+		y1 = _maxYClip;
+	if (y1 < _minYClip)
+		y1 = _minYClip;
+
+	int16 yBottom = y1;
+	int16 yTop = y1;
+	uint8 *texture = t->_texture;
```

Comment:
this buffer could be marked as const, as well as the `STexture` pointer

Reply:
Thanks! Fixed